xdsl/irdl/operations.py

```diff
@@ -914,6 +914,7 @@ class OpDef:
     or is already used by the operation, so we need to use a different name.
     """
     assembly_format: str | None = field(default=None)
+    custom_directives: dict[str, Any] = field(default_factory=dict[str, Any])
```

Comment:
My old friend import loops stopping this from being `dict[str, FormatDirective]`

Reply:
Have you tried importing if `TYPE_CHECKING`?

Reply:
and using `lambda: {}`?

Reply:
I tried a few things, one issue is that you can't even import `FormatDirective` from within `from_pyrdl` to check that the class is actually a format directive.

Reply:
you're welcome to have a play around and see if you can figure it out.